roles/lib_openshift/src/class/oc_configmap.py

```diff
@@ -0,0 +1,183 @@
+                 name,
+                 from_file,
+                 from_literal,
+                 state,
+                 namespace,
+                 kubeconfig='/etc/origin/master/admin.kubeconfig',
+                 verbose=False):
+        ''' Constructor for OpenshiftOC '''
+        super(OCConfigMap, self).__init__(namespace, kubeconfig=kubeconfig, verbose=verbose)
+        self.name = name
+        self.state = state
+        self._configmap = None
+        self._inc_configmap = None
+        self.from_file = from_file if from_file is not None else {}
+        self.from_literal = from_literal if from_literal is not None else {}
+
+    @property
+    def configmap(self):
+        if self._configmap is None:
+            self._configmap = self.get()
+
+        return self._configmap
+
+    @configmap.setter
+    def configmap(self, inc_map):
+        self._configmap = inc_map
+
+    @property
+    def inc_configmap(self):
+        if self._inc_configmap is None:
+            results = self.create(dryrun=True, output=True)
+            self._inc_configmap = results['results']
+
+        return self._inc_configmap
+
+    @inc_configmap.setter
+    def inc_configmap(self, inc_map):
+        self._inc_configmap = inc_map
+
+    def from_file_to_params(self):
+        '''return from_files in a string ready for cli'''
+        return ["--from-file={}={}".format(key, value) for key, value in self.from_file.items()]
+
+    def from_literal_to_params(self):
+        '''return from_literal in a string ready for cli'''
+        return ["--from-literal={}={}".format(key, value) for key, value in self.from_literal.items()]
+
+    def get(self):
+        '''return a configmap by name '''
+        results = self._get('configmap', self.name)
+        if results['returncode'] == 0 and results['results'][0]:
+            self.configmap = results['results'][0]
+
+        if results['returncode'] != 0 and '"{}" not found'.format(self.name) in results['stderr']:
+            results['returncode'] = 0
+
+        return results
+
+    def delete(self):
+        '''delete a configmap by name'''
+        return self._delete('configmap', self.name)
+
+    def create(self, dryrun=False, output=False):
+        '''Create a configmap
+
+           :dryrun: Product what you would have done. default: False
+           :output: Whether to parse output. default: False
+        '''
+
+        cmd = ['create', 'configmap', self.name]
+        if self.from_literal is not None:
+            cmd.extend(self.from_literal_to_params())
+
+        if self.from_file is not None:
+            cmd.extend(self.from_file_to_params())
+
+        if dryrun:
+            cmd.extend(['--dry-run', '-ojson'])
+
+        results = self.openshift_cmd(cmd, output=output)
+
+        return results
+
+    def update(self):
+        '''run update configmap '''
+        return self._replace_content('configmap', self.name, self.inc_configmap)
+
+    def needs_update(self):
```

Comment:
@stevekuznetsov,  Good catch on the `--prune` although this feels very strange to apply a prune.

>  It should be idempotent in that if the server state already matches what you need it will take no action.

The question comes in then, how does an administrator determine when things are actually changed?  With what you are proposing to switch to `oc apply` means that every task that is created, updated, deleted, or noop'd (because it exists or does not exist) would return that it changed because `oc apply` will simply  "do the right thing" and always report changed.

Here is an example:
```bash
[root@ip-172-31-60-178 ~]# oc apply -f mysql.yaml  && echo $?
deploymentconfig "mysql" configured
0
[root@ip-172-31-60-178 ~]# oc apply -f mysql.yaml  && echo $?
deploymentconfig "mysql" configured
0
[root@ip-172-31-60-178 ~]# vim mysql.yaml
[root@ip-172-31-60-178 ~]# oc apply -f mysql.yaml  && echo $?
deploymentconfig "mysql" configured
0
[root@ip-172-31-60-178 ~]# oc apply -f mysql.yaml  && echo $?
deploymentconfig "mysql" configured
0
```

Notice the vim edit in the middle.  I modified the yaml inside and as you stated, `oc apply` certainly applied it but it also applied it when it didn't need to.  This is the reason why `oc apply` is a show stopper.

I am happy to discuss the reasons why we need to see that `state: present` and `changed: True|False`.  These are imperative to the way ansible works and how operations manages clusters.